### ISSUE TITLE
optimizeOrder can return array containing nil with valid inputs

### ISSUE THREAD

**Issue by Bixx** [cayleygraph/cayley#659]
I'm running a graph query which panics on a nil pointer error. Unfortunatley it only occurs in non-trivial cases so I can't share an example. I've tracked down the error to this function:

https://github.com/cayleygraph/cayley/blob/6e9cf6c69b428fa5be7f1d12d97e184d36adea3d/graph/iterator/and_optimize.go#L156
There are several situations where the optimizeOrder function can return an array containing `[<nil>, ...]`, with valid inputs. 

For example, if `its = []`, the following loop is not run, which means `best` is not assigned a value.

https://github.com/cayleygraph/cayley/blob/6e9cf6c69b428fa5be7f1d12d97e184d36adea3d/graph/iterator/and_optimize.go#L169

However, `best` is then still appended, regardless of value:

https://github.com/cayleygraph/cayley/blob/6e9cf6c69b428fa5be7f1d12d97e184d36adea3d/graph/iterator/and_optimize.go#L204

I've tried a temporary fix by wrapping this line in a nil check:
`
if best != nil {
    out = append(out, best)
}
`

That appears to have fixed the issue for me, are there any repercussions from this?

**Comment by dennwc** [comment]
@Bixx Thanks for reporting it. I will investigate.

In case `its = []` it can actually optimize itself to Null or primary iterator if that slice does not contains primary.

**Comment by Bixx** [comment]
@dennwc, thanks for responding to this so quickly, however I am seeing the same error after updating to v0.7.0.

The stack trace is as follows:
```
[runtime error: invalid memory address or nil pointer dereference]

goroutine 194 [running]:
runtime/debug.Stack(0x7df316, 0xc4200c32a0, 0xe062a5)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x79
runtime/debug.PrintStack()
        /usr/local/go/src/runtime/debug/stack.go:16 +0x22
10.10.10.10/aaa/bbb/graph.(*Co).coTemp.func1(0xc420425592, 0xc)
        /go/src/10.10.10.10/aaa/bbb/graph/co-temp.go:103 +0x113
panic(0xd04080, 0x14232a0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.getStatsForSlice(0xc4229c22c0, 0x3, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/and_optimize.go:351 +0x80
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.materializeIts(0xc4229c22c0, 0x3, 0x4, 0x2, 0xc4229c22c0, 0x3)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/and_optimize.go:334 +0x5d
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.(*And).Optimize(0xc421c35dc0, 0x1447bc0, 0xc42209a8c0, 0xc42209abe0)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/and_optimize.go:74 +0x29e
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.(*Unique).Optimize(0xc42209aaa0, 0x0, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/unique.go:116 +0x38
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.optimizeSubIterators(0xc422fb0000, 0x2, 0x14, 0x30, 0xc421350f30, 0xc422ac13e8)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/and_optimize.go:276 +0x8d
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.(*Or).Optimize(0xc421400930, 0xc421350f30, 0x2, 0xc422ac1418)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/or.go:260 +0x68
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator.(*Unique).Optimize(0xc42209ab40, 0x1447b01, 0xc42209ab40, 0xc422ac1468)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterator/unique.go:116 +0x38
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph.(*IterateChain).start(0xc4203eff90)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterate.go:65 +0x2b8
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph.(*IterateChain).Each(0xc4203eff90, 0xc422ac1570, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterate.go:118 +0x41
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph.(*IterateChain).EachValue(0xc4203eff90, 0x0, 0x0, 0xc422ac15c8, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterate.go:285 +0x93
10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph.(*IterateChain).AllValues(0xc4203eff90, 0x0, 0x0, 0xc4203eff90, 0xc421f4db40, 0xc421f4db00, 0x0, 0xc421f4d700)
        /go/src/10.10.10.10/aaa/bbb/vendor/github.com/cayleygraph/cayley/graph/iterate.go:311 +0x87
10.10.10.10/aaa/bbb/graph.(*Co).findLinked(0xc420379a20, 0xc420425592, 0xc, 0xc42250c000, 0x63, 0x80, 0xc4202fa1a0, 0xc42078a670, 0x19, 0x1, ...)
        /go/src/10.10.10.10/aaa/bbb/graph/co-temp.go:423 +0x1c15
10.10.10.10/aaa/bbb/graph.(*Co).coTemp(0xc420379a20, 0xc42032d3c0, 0xd, 0xc420425592, 0xc, 0xc4202fa1a0, 0x0, 0x0, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/graph/co-temp.go:132 +0x5f1
10.10.10.10/aaa/bbb/graph.(*Co).Co(0xc420379a20, 0xc42032d3c0, 0xd, 0xc420425592, 0xc, 0x0, 0x0, 0x0, 0x0)
        /go/src/10.10.10.10/aaa/bbb/graph/co-temp.go:80 +0x1d7
10.10.10.10/aaa/bbb/pipeline.(*Pipeline).initEn.func1(0xc4203ded20)
        /go/src/10.10.10.10/aaa/bbb/pipeline/pipeline.go:227 +0x350
created by 10.10.10.10/aaa/bbb/pipeline.(*Pipeline).initEn
        /go/src/10.10.10.10/aaa/bbb/pipeline/pipeline.go:250 +0x33a

```

**Comment by dennwc** [comment]
@Bixx Can you please verify that it's fixed upstream? (128d78457ad9c94383d76164e815a1c0301287a5)

**Comment by Bixx** [comment]
@dennwc that works for me. I've been patching with this fix for some time now and haven't encountered any further issues. Thanks very much.
